osvr_tracked_device.h

```diff
@@ -340,107 +331,49 @@ bool OSVRTrackedDevice::IsDisplayRealDisplay()
-        *height = m_DisplayConfiguration->getDisplayHeight() / 2;
-        *x = 0;
-        *y = (vr::Eye_Left == eye) ? 0 : *height;
-        break;
-    case OSVRDisplayConfiguration::FULL_SCREEN:
-        *width = m_DisplayConfiguration->getDisplayWidth();
-        *height = m_DisplayConfiguration->getDisplayHeight();
-        *x = 0;
-        *y = 0;
-        break;
-    default:
-        *width = m_DisplayConfiguration->getDisplayWidth();
-        *height = m_DisplayConfiguration->getDisplayHeight();
-        *x = 0;
-        *y = 0;
-        std::cerr << "ERROR: Unexpected display mode type: " << m_DisplayConfiguration->getDisplayMode() << ".\n";
-    }
+    osvr::clientkit::RelativeViewport viewPort = m_DisplayConfig.getViewer(0).getEye(eye).getSurface(0).getRelativeViewport();
+    *x = viewPort.left;
+    *y = viewPort.bottom;
+    *width = viewPort.width;
+    *height = viewPort.height;
 }
 
 void OSVRTrackedDevice::GetProjectionRaw(vr::Hmd_Eye eye, float* left, float* right, float* top, float* bottom)
 {
-    /// @todo Use OSVR API to obtain this data
-
-    // Projection matrix centered between the eyes
-    const double z_near = 0.1;
-    const double z_far = 100.0;
-    const Eigen::Matrix4d center = make_projection_matrix(m_DisplayConfiguration->getHorizontalFOVRadians(), m_DisplayConfiguration->getFOVAspectRatio(), z_near, z_far);
-
-    // Translate projection matrix to left or right eye
-    double eye_translation = m_DisplayConfiguration->getIPDMeters() / 2.0;
-    if (vr::Eye_Left == eye) {
-        eye_translation *= -1.0;
-    }
-    const Eigen::Affine3d translation(Eigen::Translation3d(Eigen::Vector3d(eye_translation, 0.0, 0.0)));
-
-    const Eigen::Matrix4d eye_matrix = translation * center;
-
     // Reference: https://github.com/ValveSoftware/openvr/wiki/IVRSystem::GetProjectionRaw
-    // This is similar to glFrustum but not quite the same!
-    // idx = 1 / (right - left)
-    // idy = 1 / (bottom - top)
-    // idz = 1 / (far - near)
-    // sx = right + left
-    // sy = bottom + top
-    // eye(0, 0) = 2 * idx = 2 / (right - left)
-    // eye(1, 1) = 2 * idy = 2 / (bottom - top)
-    // eye(0, 2) = sx * idx = (right + left) / (right - left)
-    // eye(1, 2) = sy * idy = (bottom + top) / (bottom - top)
-    // eye(2, 2) = -far * idz = -far / (far - near)
-    // eye(2, 3) = -(far * near * idz) = -(far * near) / (far - near)
-    // eye(3, 2) = -1
-
-    const double dx = 2.0 / eye_matrix(0, 0); // (right - left)
-    const double sx = eye_matrix(0, 2) * dx; // (right + left)
-    *right = static_cast<float>((sx + dx) / 2.0); // right
-    *left = static_cast<float>(sx - *right);
-
-    const double dy = 2.0 / eye_matrix(1, 1); // (bottom - top)
-    const double sy = eye_matrix(1, 2) * dy; // (bottom + top)
-    *bottom = static_cast<float>((sy + dy) / 2.0); // bottom
-    *top = static_cast<float>(sy - *bottom);
+    // SteamVR expects top and bottom to be swapped!
```

Comment:
Ah, I see. I think if you actually use the OSVR projection matrix functions, you can specify flags that would perform the transformation between those two formats (I did sit down and figure out all the different ways one might want their projection matrix....), but since they want top/bottom/left/right, they are getting them before those options become available.

Although, that wasn't the element I thought it was, it's 3,4 and 4,4 that are the weird ones, and that's more than a sign flip so I don't think it's done elsewhere - could it be a difference between "coordinate for the clipping plane" (what that doc you linked to me says) and "distance to the clipping plane" (what I thought the spec for glFrustum and our API was) and that switching those two bounds just gets closer to a correct answer?